### ISSUE TITLE
Consider separating core Hazelcast JAR into smaller units

### ISSUE THREAD

**Issue by vpavic** [hazelcast/hazelcast#7448]
Using Hazelcast Java Client currently requires server libraries on the classpath as well. This is not the optimal solution both in terms of size of dependencies that have to added to a project, and from perspective of other libraries/frameworks which provide added value on top of Hazelcast functionalities.

Separation into smaller units that users would pick according to their needs would make sense, and would also address aforementioned concerns.

See spring-projects/spring-boot#4918 for motivation and the original discussion.

**Comment by pveentjer** [comment]
We are currently doing a lot of work for better modularization. I don't know if remving the serverside client classes is part of this. @asimarslan do you know if these classes are going to be moved out of the core jar and into a separate jar?

**Comment by asimarslan** [comment]
In order to achieve single jar client goal we need to split hazelcast itself into multiple modules like serialization, api, util/common etc. Until then we cannot move those classes from core.

**Comment by pveentjer** [comment]
Splitting up in different jars isn't a breaking change. This can always be done in a minor release So moving it to 4.1

**Comment by vpavic** [comment]
@pveentjer I don't see how this is not a breaking change - once this change is made, it will have a direct impact on what user's declared Hazelcast dependencies bring to the classpath.

For instance, at the moment it is possible to depend only on `com.hazelcast:hazelcast-client` and use Hazelcast server since this dependency transitively pulls in `com.hazelcast:hazelcast`. When this change is made, this will no longer be possible therefore it is a breaking change.

**Comment by mmedenjak** [comment]
As of Hazelcast 4.0, the client is now part of the core jar. So you don't need an extra JAR if you want to use clients.

**Comment by vpavic** [comment]
> As of Hazelcast 4.0, the client is now part of the core jar.

Unfortunately, this is the move in opposite direction of the one originally proposed here. What's actually for this issue, since it's still in the `4.1` milestone?

Ping @snicoll - this is something to be aware of when Spring Boot picks up Hazelcast `4.0`.

**Comment by snicoll** [comment]
Thanks for the heads-up @vpavic.

**Comment by mmedenjak** [comment]
For now, we have merged client into the core JAR as that was more of an issue for the users. We'll keep this issue open in case there is more interest in splitting it back again.

**Comment by snicoll** [comment]
I can see that this issue was assigned to the 4.1 backlog. Are you reconsidering [the change you made in v4](https://github.com/hazelcast/hazelcast/issues/7448#issuecomment-546338038) and will you split those jars after all?

I am asking because it makes the upgrade to v4 in Spring Boot quite difficult as we've mentioned previously here. Having both in the same jar will not let us figure out if the user has opt-in for the client support. If we don't have that information, then the user has to provide that to use (in the form of a configuration property I suspect).

The problem with the configuration property is that we can architecture it in two ways IMO:

* It doesn't have a default and you have to configure it for the Hazelcast auto-configuration to kick in. This is a breaking change for users and I don't see that as an improvement, really
* It does have a default, which means we have to chose what takes precedence between the embedded and c lient mode. Do we know for sure what that order should be?

**Comment by vpavic** [comment]
As a long-time Hazelcast user, it always struck as weird to have server components needed on the classpath when I'm simply connecting to a remote cluster. The `hazelcast.jar` is over 11 MB big and it's not reasonable to have to pull in that big of a dependency in some simple application just to be able to connect to a remote Hazelcast cluster.

The change made in `4.0` packaging was a step back IMO, because now there's the same weirdness of having client components present on the classpath when I'm building a Hazelcast (embedded) server.

As mentioned several times, this is also inconvenient for anyone trying to build an added value on top of Hazelcast, such as Spring Boot's auto-configuration. So I'm still strongly in favour of having a clean separation as indicated in this issue's opening comment.

**Comment by mmedenjak** [comment]
@snicoll no, we're not reconsidering splitting it in the near future. The issue has been moved to backlog for now and we _probably_ won't make any changes until the next major version because of backwards compatibility guarantees.

Still, I understand how this does not fare well with your use case and I think it is worth revisiting this decision at some point in the future. Inviting @sancar and @asimarslan to share the exact reasons why we decided to go in this direction and if they foresee issues with splitting the code into multiple JARs.

**Comment by mmedenjak** [comment]
@vpavic @snicoll Not that it's going to justify the decision for your case but I can give you some context about why we decided to merge the client and core hazelcast JAR for Hazelcast 4.0. 

As you might know, to use clients in 3.x, you needed to have both the `hazelcast-client.jar` and the `hazelcast.jar` on the classpath. During 4.0 development, we determined that the `hazelcast.jar` dominated in the size and that the `hazelcast-client.jar` didn't actually add a lot (from https://mvnrepository.com/artifact/com.hazelcast/hazelcast-client/3.12, it looks like it's about an additional 928 KB). 

We then decided to merge the client code into the core JAR because:
- users using clients had to download both JARs anyway
- it was simpler for users using clients to start using hazelcast, only 1 jar is needed
- the client code only made a small difference in size for users not using clients

So, when we're speaking about JAR sizes, I don't see much difference between the 3.x state of things and 4.0 state of things. In addition to this, these are the JAR sizes of `hazelcast.jar` in 3.12 (without client code) and 4.0 (with client code):
```
https://mvnrepository.com/artifact/com.hazelcast/hazelcast/3.12 -> 11.3 MB
https://mvnrepository.com/artifact/com.hazelcast/hazelcast-all/3.12 -> 12.7 MB
https://mvnrepository.com/artifact/com.hazelcast/hazelcast/4.0 -> 10.7 MB
https://mvnrepository.com/artifact/com.hazelcast/hazelcast-all/4.0 -> 11.1 MB
```

So, the only issue I see with our decision is that it is harder for spring boot to determine if the user now wants to start a client or a member. Yes, I admit this had made the situation worse for you and as far as I hear, the only suggestion is adding additional configuration.

I also admit that in the long run, it may be preferable to split the JAR into several smaller ones (like `hazelcast-core.jar` (having the infrastructure), `hazelcast-member.jar` (having only the member-side code) and `hazelcast-client.jar` (having only the client-side code) but this requires a lot more engineering effort than what we were able to give during 4.0. Also, more importantly, since clients reuse a lot of the member-side code, it _may_ happen that we end up with a similar state as 3.x, that the core/common/infrastructure JAR dwarfs the other two JARs in size (e.g. `hazelcast-core.jar` ~10MB, `hazelcast-member.jar` <1MB, `hazelcast-core.jar` <1MB), thus making the benefits of this split  questionable.

**Comment by vpavic** [comment]
> So, when we're speaking about JAR sizes, I don't see much difference between the 3.x state of things and 4.0 state of things.

I didn't suggest that the size is not a concern introduced by `4.0`. It was actually among the concerns expressed when this issue [was raised](https://github.com/hazelcast/hazelcast/issues/7448#issue-130674056), over 4 years ago.

Over 11 MB worth of dependencies is a tough pill to swallow for e.g. some embedded client operating in resource-constrained environment that's trying to connect to Hazelcast cluster.

**Comment by rmannibucau** [comment]
Up, any way it is sorted out soon? core is super big now and moreover embeds dependencies it is unlikely to desire - or more exactly you want to control in your app to be able to apply a security policy on it whereas with relocations it is requires a lot of work for no real gain (like jackson to give just one example).

**Comment by mmedenjak** [comment]
I'm not aware of any plans in the immediate future. Adding @burakcelebi to raise awareness.

**Comment by mmedenjak** [comment]
Also, I think we're talking about two things here:
- first is the size of the client JAR, which I believe needs to be addressed
- second is the relocation (shading) of the dependencies, which was not mentioned up until now. Unlike the first, I don't consider this as an issue. Yes, it poses a problem for security policies but on the other hand not shading the dependencies might cause dependency conflicts when running alongside many other popular libraries on the classpath that require different versions of the shared dependencies, i.e. dependency hell.

**Comment by rmannibucau** [comment]
@mmedenjak 

1. agree + the "server" one as well (not only the client), this is how it is linked to 2
2. dependency hell is quite trivial to fix if you put a light SPI, typical example will be that you can drop jackson in favor of Jsonb kind of SPI (maybe just jsonp) and the impl can be a SPI like which would enable dependency hell to be solved by switching either the impl or just the version/vendor. The SPI surface is way easier to manage in time while you don't leak any vendor API and you never get any dependency hell for the same reason relocation works today (it is your internal API). So at the end it only has the pitfalls to hide vulnerabilities and reduce the auditing capabilities.

**Comment by mmedenjak** [comment]
We'll consider it but one thing to note is that we also use some libraries (e.g. Apache Calcite) which don't have an SPI and depend on other libraries (e.g. jackson-core, guava, slf4j, ...) directly.

**Comment by rmannibucau** [comment]
@mmedenjak yes but if you put calcite behind a hz PSI, you can use jsqlparser as impl for example (this is what HerdDB does for example). So the SPI should be in front of optional and "big" subtree IMHO.

**Comment by mmedenjak** [comment]
We can try it. We don't just use calcite for parsing so I'm not sure how feasible this approach will be.

**Comment by rmannibucau** [comment]
@mmedenjak yes it is not trivial but from what I saw in HerdDB, hooking itself in calcite is often as much work as doing the full impl yourself. Only real benefit are coming from optimizations related to joins and nullable but hz has the model for that so should be able to do it. Also note that for this particular issue, having calcite as an external module only is fine while the main/default bundle does not bring it.

**Comment by sancar** [comment]
Although we think this is critical, we could not find manpower to fix this on 5.0. Moving this to backlog for now. We will assign a milestone when we plan for a specific one.
